File: Version-Python/models/customer.py
from uuid import uuid4
from .person import Person

class Customer(Person):
    def __init__(self, name, contact_info, gender):
        super().__init__(name, contact_info, gender)
        self.order_history = []

    def add_customer(self, supermarket):
        supermarket.customers.append(self)
        print(f'Customer "{self.name}" added to supermarket "{supermarket.name}".')

    def get_customer_infos(self):
        print(f"""
        Customer Infos:
        -----------------
        
        ID: {self.id}
        Name: {self.name}
        Contact Info: {self.contact_info}
        Sexe: {self.gender}
        Order History: {len(self.order_history)}
        """)

    def delete_customer(self, supermarket):
        supermarket.customers = [c for c in supermarket.customers if c.id != self.id]
        print(f'Customer "{self.name}" removed from supermarket "{supermarket.name}".')

    def place_order(self, supermarket, product_id, quantity):
        if not all([supermarket, product_id, quantity]):
            raise ValueError("All required information should be provided!")

        product = next((p for p in supermarket.products if p.id == product_id), None)
        if not product:
            raise ValueError("Product not found!")
        
        if product.quantity < quantity:
            raise ValueError("Insufficient stock!")

        product.quantity -= quantity
        
        order = {
            'id': str(uuid4()),
            'customer_id': self.id,
            'product_id': product_id,
            'quantity': quantity,
            'date': datetime.now()
        }
        
        self.order_history.append(order['id'])
        print(f'{quantity} Products of ID "{product_id}" has been successful supplied!')

File: Version-JavaScript/models/Customer.js
import Person from './Person.js';

// Class to Manage Customer who is above all a Person
class Customer extends Person {
    constructor(name, contactInfo, gender) {
        super(name, contactInfo, gender);
        this.orderHistory = [];
    }

	// Add a Customer to inventory of supermaket
    addCustomer(supermarket) {
        supermarket.customers.push(this);
        console.log(`Customer "${this.name}" added to supermarket "${supermarket.name}".`);
    }

	// Method to get Customer informations
	getCustomerInfos() {
		console.log(`
		Customer Infos:
		-----------------
		
		ID: ${this.id}
		Name: ${this.name}
		Contact Info: ${this.contactInfo}
		Sexe: ${this.gender}
		Order History: ${this.orderHistory.length}
		`);
	}
	
	// Method to delete a Customer from a suppermaket
    deleteCustomer(supermarket) {
        supermarket.customers = supermarket.customers.filter(customer => customer.id !== this.id);
        console.log(`Customer "${this.name}" removed from supermarket "${supermarket.name}".`);
    };
    
    // Method to place an order depending of the productID and the quantity
	placeOrder(supermaket, productID, quantity) {
		if(!supermaket || !productID || !quantity) {
			throw new Error("All required informations should be provided !!!");
		}
		
		const product = supermaket.products.find(p => p.id === productID);
        if (!product) throw new Error("Product not found!");
        
        
        if (product.stock < quantity) throw new Error("Insufficient stock!");

        product.stock -= quantity;
        
		 const order = {
            id: uuidv4(),
            customerId: this.id,
            productID,
            quantity,
            date: new Date(),
        };
		
		this.orderHistory.push(order.id);
		
		// Return success message after all
		console.log(`${quantity} Products of ID "${productID}" has been successful supplied !`);
	};
}

export default Customer;

// Test
const customer = new Customer("Customer B", "customer@example.com", "Female");
// customer.getCustomerInfos();
